pkg/controller/common/name/name.go

```diff
@@ -40,33 +70,55 @@ func (n Namer) WithDefaultSuffixes(defaultSuffixes ...string) Namer {
 // Suffix a resource name.
 //
 // Panics if the suffix exceeds the suffix limits.
-// Trims the name if the concatenated result would exceed the limits.
 func (n Namer) Suffix(ownerName string, suffixes ...string) string {
+	suffixedName, err := n.SafeSuffix(ownerName, suffixes...)
+	if err != nil {
+		// we should never encounter an error at this point because the names should have been validated
+		log.Error(err, "Invalid name. This could prevent the operator from functioning correctly", "name", suffixedName)
+	}
+
+	return suffixedName
+}
+
+// SafeSuffix attempts to generate a suffixed name, returning an error if the generated name is unacceptable.
+func (n Namer) SafeSuffix(ownerName string, suffixes ...string) (string, error) {
 	var suffixBuilder strings.Builder
+	var err error
+
 	for _, s := range n.DefaultSuffixes {
-		suffixBuilder.WriteString("-") // #nosec G104
-		suffixBuilder.WriteString(s)   // #nosec G104
+		suffixBuilder.WriteString("-")
+		suffixBuilder.WriteString(s)
 	}
 	for _, s := range suffixes {
-		suffixBuilder.WriteString("-") // #nosec G104
-		suffixBuilder.WriteString(s)   // #nosec G104
+		suffixBuilder.WriteString("-")
+		suffixBuilder.WriteString(s)
 	}
 
 	suffix := suffixBuilder.String()
 
 	// This should never happen because we control all the suffixes!
 	if len(suffix) > n.MaxSuffixLength {
-		panic(fmt.Errorf("suffix should not exceed %d characters: got %s", n.MaxSuffixLength, suffix))
+		err = multierror.Append(err, newNameLengthError("suffix exceeds max length", n.MaxSuffixLength, suffix))
+		suffix = truncate(suffix, n.MaxSuffixLength)
 	}
 
-	// This should never happen because the ownerName length should have been validated.
-	// Trim the ownerName and log an error as fallback.
-	maxPrefixLength := MaxNameLength - len(suffix)
+	maxPrefixLength := validation.LabelValueMaxLength - len(suffix)
 	if len(ownerName) > maxPrefixLength {
-		log.Error(fmt.Errorf("ownerName should not exceed %d characters: got %s", maxPrefixLength, ownerName),
-			"Failed to suffix resource")
-		ownerName = ownerName[:maxPrefixLength]
+		err = multierror.Append(err, newNameLengthError("owner name exceeds max length", maxPrefixLength, ownerName))
+		ownerName = truncate(ownerName, maxPrefixLength)
+	}
+
+	return stringsutil.Concat(ownerName, suffix), err
+}
+
+func truncate(s string, length int) string {
+	var b strings.Builder
+	for _, r := range s {
+		if b.Len()+utf8.RuneLen(r) > length {
```

Comment:
Do we expect anything other than single byte characters? Should we validate that somewhere?

Reply:
Kubernetes' own validation checks will not let multi-byte names to be submitted at the moment but I prefer not to rely on that assumption.

Reply:
Got it. Should we check it in the Validate function then? It seems here we truncate based on (potential) multibyte length, but other than that we allow it and rely on k8s own validation.

Reply:
So, to provide some context here, the validation webhook simply transitions the object to the `INVALID` state when the custom validation checks fail. Trying to delete the object causes a problem because the finalizers start running and then start to trip over the invalid names -- making it impossible to delete. This is why I introduced the truncate function so that the object can be deleted even when it is invalid. It should never be used in the happy path and therefore whether the string is multi-byte or not does not really matter.